velox/expression/Expr.cpp

```diff
@@ -267,6 +267,104 @@ void Expr::releaseInputValues(EvalCtx& evalCtx) {
   inputValues_.clear();
 }
 
+namespace {
+
+void ensureErrorsSize(EvalCtx::ErrorVector& errors, int32_t size) {
+  if (size > errors.size()) {
+    int32_t oldSize = errors.size();
+    errors.resize(size);
+    for (auto i = oldSize; i < size; ++i) {
+      errors.setNull(i, true);
+    }
+  }
+}
+
+// Adds the errors from 'errors' to 'result'.
+void addErrors(
+    const EvalCtx::ErrorVector& errors,
+    EvalCtx::ErrorVector& result) {
+  ensureErrorsSize(result, errors.size());
+  bits::forEachBit(
+      errors.rawNulls(), 0, errors.size(), bits::kNotNull, [&](auto row) {
+        result.set(row, errors.valueAt(row));
+      });
+}
+
+void rethrowFirstError(const EvalCtx::ErrorVectorPtr& errors) {
+  bits::forEachBit(
+      errors->rawNulls(), 0, errors->size(), bits::kNotNull, [&](auto row) {
+        auto exceptionPtr =
+            std::static_pointer_cast<std::exception_ptr>(errors->valueAt(row));
+        std::rethrow_exception(*exceptionPtr);
+      });
+}
+
+void mergeOrThrowArgumentErrors(
+    EvalCtx::ErrorVectorPtr& errors,
+    EvalCtx::ErrorVectorPtr& argumentErrors,
+    EvalCtx& context) {
+  if (argumentErrors) {
+    if (context.throwOnError()) {
+      rethrowFirstError(argumentErrors);
+    }
+    if (errors) {
+      addErrors(*argumentErrors, *errors);
+    }
+    context.swapErrors(argumentErrors);
+  } else {
+    context.swapErrors(errors);
+  }
+}
+
+template <typename GetRemainingRows>
+void processNullsAndErrorsFromArgument(
+    const uint64_t* flatNulls,
+    GetRemainingRows getRemainingRows,
+    EvalCtx::ErrorVectorPtr& errors,
+    EvalCtx::ErrorVectorPtr& argumentErrors,
+    EvalCtx& context) {
+  // A null with no error deselects the row and clears a possible error. An
+  // error adds itself to argument errors.
+  if (context.errors()) {
```

Comment:
Are context.errors() represent errors that occurred during evaluation of a single argument?

What does flatNulls represent?